### ISSUE TITLE
How do I set Project attributes?

### ISSUE THREAD

**Issue by mkosmala** [zooniverse/Panoptes#714]
Do I follow "Edit a Single Project" in Projects? If so, do I need to pull the entire JSON description for a project, edit it, and then send it back? Or can I just send the attribute I want to set? I've tried to set the science case of a Project using the latter technique without success.

Related, how do I set the icon and background image?

**Comment by mkosmala** [comment]
I am, for example doing this:
```
# Adding science case
URL: https://panoptes-staging.zooniverse.org/api/projects/182

Payload:
        {
            "projects:" {
                "science_case": "default science case"
            }
        }
```
And getting error:
```
Status code: 400
Response text: HTTP/1.1 500 Internal Server Error
```

**Comment by mkosmala** [comment]
Likewise:
```
# Adding introduction
URL: https://panoptes-staging.zooniverse.org/api/project_contents/182

Payload:
        {
            "project_contents:" {
                "introduction": "default introduction"
            }
        }
```
yields
```
Status code: 400
Response text: HTTP/1.1 500 Internal Server Error
```

**Comment by camallen** [comment]
In response to:
>Do I follow "Edit a Single Project" in Projects? If so, do I need to pull the entire JSON description for a project, edit it, and then send it back? Or can I just send the attribute I want to set? I've tried to set the science case of a Project using the latter technique without success.

So we don't support the HTTP Patch operation to do partial updates. We do support PUT(update) and POST(create) operations both to my knowledge require a full representation of the resource unless otherwise documented. 

Updating the project contents you should use the full representation for both Create (POST) and Update (PUT). 

All our error messages should have more details in the response body, we've tried to make these as informative about the error as we can. Your first port of call when getting an error from the API should be the status code, then the response body.

FYI - You can read more about our JSON api implementation spec here, http://jsonapi.org/format/#crud-updating the docs have been evolving and our interpretation may be a little out of sync.

**Comment by mkosmala** [comment]
Am I right in assuming that when I create a Project the ProjectContents are created automatically? If so, what's the best way to put in things like the science_case and introduction? Create a Project, retrieve the ProjectContents, and then update the ProjectContents?

**Comment by camallen** [comment]
Yes - https://github.com/zooniverse/Panoptes/blob/master/app/controllers/api/v1/projects_controller.rb#L52

On create via these fields are allowed: https://github.com/zooniverse/Panoptes/blob/master/app/schemas/project_create_schema.rb

Update these fields: https://github.com/zooniverse/Panoptes/blob/master/app/schemas/project_update_schema.rb

**Comment by camallen** [comment]
So it looks like you can't edit the primary language `project_contents` directly. You can do so via an update on the project though see [here](https://github.com/zooniverse/Panoptes/blob/master/spec/controllers/api/v1/projects_controller_spec.rb#L353) and link the content fields up to [here](https://github.com/zooniverse/Panoptes/blob/master/app/schemas/project_update_schema.rb). That should allow you to edit the project contents.

**Comment by edpaget** [comment]
The reason you can't edit the primary project_contents directly is to prevent users with the 'translator' role from editing the base page. @camallen do you think we should work something out to make sure the primary content is editable through the /project_contents routes?

**Comment by camallen** [comment]
I think providing better error messaging around why it failed would have alleviated the issue. Any reason why we can't reject `non-owner` role access when editing primary project content?

**Comment by edpaget** [comment]
I'm not sure this is going to be possible without really overhauling how permissions work. I think this would generate an unacceptably slow SQL query. I would like it to eventually work, but I'm going to move it to Future plans for now.
